### PR TITLE
agent: Support fetching LFC metrics (but don't use them yet)

### PR DESCRIPTION
**Pull request by sharnoff**
Probably best to review each commit separately.

Most of the commits here are just repositioning to make the final commit, which adds LFC metrics collection, as simple as possible.

~~Builds on #892, must not be merged before then.~~

### CODE REVIEW

File: pkg/api/vminfo.go
Comment (Omrigan) [inline pkg/api/vminfo.go:374]: Did you mean deep copy?

Also, WDYT about https://go.dev/play/p/wDUnmOqCSqC ?

File: pkg/agent/runner.go
Comment (Omrigan) [inline pkg/agent/runner.go:271]: ctx2, logger2?

Otherwise, `r, c, l` is confusing IMO

File: pkg/agent/runner.go
Comment (Omrigan) [inline pkg/agent/runner.go:379]: Can you have it as a part of an interface on C? 

Also, C is metrics names, right, maybe should be `Names`?

File: pkg/agent/runner.go
Comment (Omrigan) [inline pkg/agent/runner.go:380]: Can you create empty M, and drop this arg?

**Review by Omrigan — COMMENTED** [review]
This takes a little bit of time to review for me, as I am not used to this many generics usages. Usually, I would advocate against it in Go, relaying on dynamic dispatch/code duplication. But since we have Rust in the picture, maybe I should get used to it.

LMK if you come up the a way to make it less generic

**Comment by skyzh** [comment]
Discussion from the 1-1 meeting: scraping sql_exporter from the autoscaling-agent is the right way to go. I'll add a separate sql_exporter solely for autoscaling metrics. For now, we can just scrape the current one with a reasonable interval.

File: pkg/api/vminfo.go
Comment (sharnoff) [reply pkg/api/vminfo.go:374]: Hm... control plane uses https://github.com/samber/lo's [`lo.ToPtr`](https://pkg.go.dev/github.com/samber/lo#ToPtr) to fulfill this purpose - see e.g. here:

https://github.com/neondatabase/cloud/blob/2b0562e85d4586ad62a8c7af84fbc0af647391f0/goapp/internal/operationsv2/provisioner/provisioner_k8s_common.go#L289-L293

We also define a similar helper function in `pkg/agent/core/state_test.go`:

https://github.com/neondatabase/autoscaling/blob/4d83eb6efe45b08ea8d8a1c93d69013a1a23b549/pkg/agent/core/state_test.go#L234-L236

I'm hesitant to introduce a temporary function for just this location, but I'd be happy to open a follow-up PR to tackle this repo-wide ?

WDYT?

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go:271]: That's fair. All of the other functions in that area use the same naming — I'm hesitant to change these without changing those as well. The reason they're like that currently is because I thought shadowing would be confusing, and it'd be too verbose to write out `ctx2` and `logger2` in all of the usages.

What do you think?

File: pkg/agent/core/metrics.go
Comment (sharnoff) [reply pkg/agent/core/metrics.go:154]: Makes sense, done. See (0af52ef29d80836bfbb698e4bc449e86b6cbe550) and (fe8737b19026effd9829abab18e5e8c94b299f6a).

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go:379]: > Can you have it as a part of an interface on C?

Not quite sure what you're referring to here — is it `kind`?

> Also, C is metrics names, right, maybe should be `Names`?

I was originally thinking we'd want some per-VM configuration for metrics for feature-gating new behavior, but upon reflection, I think implementing that properly is actually more effort than it's worth.

At the same time... I think the semantics are clearer if the interface defines it as "config" for metrics parsing, and then something external provides that "config" - even if it's always just the metric names in practice.

For now, I've added some comments to `FromPrometheus` (f47097a984e94f8b704a86f01d9c987e2dd1acd7). LMK what you think.

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go:380]: Unfortunately `M` is basically always a pointer, which makes this difficult.

We _can_ solve the issue by further convoluting the generics — maybe it's worthwhile, maybe it's not. That would look something like:

```diff
diff --git a/pkg/agent/runner.go b/pkg/agent/runner.go
index 52cecb0..5216a3c 100644
--- a/pkg/agent/runner.go
+++ b/pkg/agent/runner.go
@@ -256,7 +256,6 @@ func (r *Runner) Run(ctx context.Context, logger *zap.Logger, vmInfoUpdated util
 			c,
 			l,
 			"system",
-			func() *core.SystemMetrics { return new(core.SystemMetrics) },
 			func() bool { return true },
 			r.global.config.Metrics.System,
 			func(metrics *core.SystemMetrics, withLock func()) {
@@ -270,7 +269,6 @@ func (r *Runner) Run(ctx context.Context, logger *zap.Logger, vmInfoUpdated util
 			c,
 			l,
 			"LFC",
-			func() *core.LFCMetrics { return new(core.LFCMetrics) },
 			func() bool {
 				scalingConfig := r.global.config.Scaling.DefaultConfig.WithOverrides(getVmInfo().Config.ScalingConfig)
 				return *scalingConfig.EnableLFCMetrics // guaranteed non-nil as a required field.
@@ -369,15 +367,19 @@ func (r *Runner) spawnBackgroundWorker(ctx context.Context, logger *zap.Logger,
 	}()
 }
 
+type metricsPointer[C any, V any] interface {
+	core.FromPrometheus[C]
+	~*V
+}
+
 // getMetricsLoop repeatedly attempts to fetch metrics from the VM
 //
 // Every time metrics are successfully fetched, the value is recorded with newMetrics.
-func getMetricsLoop[M core.FromPrometheus[C], C any](
+func getMetricsLoop[M metricsPointer[C, V], C any, V any](
 	r *Runner,
 	ctx context.Context,
 	logger *zap.Logger,
 	kind string,
-	emptyMetrics func() M,
 	isActive func() bool,
 	config MetricsSourceConfig[C],
 	newMetrics func(metrics M, withLock func()),
@@ -418,14 +420,15 @@ func getMetricsLoop[M core.FromPrometheus[C], C any](
 			}
 			lastActive = true
 
-			metrics := emptyMetrics() // dependency-injected because M is usually a pointer, and otherwise the generics get too annoying.
-			err := doMetricsRequest[M, C](r, ctx, logger, timeout, metrics, config.Names)
+			var metrics V
+
+			err := doMetricsRequest[M, C](r, ctx, logger, timeout, M(&metrics), config.Names)
 			if err != nil {
 				logger.Error("Error making metrics request", zap.Error(err))
 				goto next
 			}
 
-			newMetrics(metrics, func() {
+			newMetrics(M(&metrics), func() {
 				logger.Info("Updated metrics", zap.Any("metrics", metrics))
 			})
 		}
```

The shenanigans at the bottom with repeated `M(&metrics)` _could_ get cleaner with https://github.com/samber/lo, because we could instead construct an `M` in a single line :sweat_smile:

File: pkg/api/vminfo.go
Comment (Omrigan) [reply pkg/api/vminfo.go:374]: > en a follow-up PR to tackle this repo-wide ?

Sure, when you have the time.

**Review by Omrigan — COMMENTED** [review]
I think mostly nits are left, we should discuss them verbally

**Comment by sharnoff** [comment]
Discussion with @Omrigan : It's fine to just hard-code the metric names, and that makes it a lot simpler. Some discussion about whether to encapsulate the callbacks for metrics collection into a single struct with `func(...)` fields or an interface.

File: pkg/agent/runner.go
Comment (Omrigan) [reply pkg/agent/runner.go:281]: Thought about this a bit, and I guess creating another type would require passing some dependencies to it, which will make it more verbose. 

So defining a struct and passing callbacks through it, so that they are named, and we don't have to guess what :question:  `func() bool { return true }` :question:  does.

File: pkg/api/vminfo.go
Comment (sharnoff) [reply pkg/api/vminfo.go:374]: PR here: #946

File: pkg/api/vminfo.go
Comment (sharnoff) [reply pkg/api/vminfo.go:406]: Fixed in #948 bba199cbc7fa33499285f17effefa8ba1ec30946

File: pkg/agent/config.go
Comment (sharnoff) [reply pkg/agent/config.go]: Resolved this with [`4b06d09`..`0bb2b59`](https://github.com/neondatabase/autoscaling/compare/4b06d09cc3545f5f386cf1d8fe7bd90165ff7a75..0bb2b59f8326a70c6438a1623f98d69ae4c4f34e), as discussed :)

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go]: Implemented this in 5dc98a95675fcb6573278834a688b123eda89149

File: pkg/agent/config.go
Comment (sharnoff) [reply pkg/agent/config.go]: I think this is no longer needed now that metric names have been moved into the code?

Resolving; feel free to reopen if I misjudged

File: pkg/agent/config.go
Comment (sharnoff) [reply pkg/agent/config.go]: Yep - implemented in 86cb9f4a277a5357f95920a82d25b074eef3e60a and c1ae8d3b1ed5c2badcf0d3873b939b6be70137a9 (kept separate so that they can be properly squashed later).

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go]: I think 5dc98a95675fcb6573278834a688b123eda89149 satisfies this (in addition to the changes from [`4b06d09`..`0bb2b59`](https://github.com/neondatabase/autoscaling/compare/4b06d09cc3545f5f386cf1d8fe7bd90165ff7a75..0bb2b59f8326a70c6438a1623f98d69ae4c4f34e) ?

**Comment by sharnoff** [comment]
New changes:

1. Move metric names into the code (rather than config) and remove the generics it required. Diff here: [`4b06d09`..`0bb2b59`](https://github.com/neondatabase/autoscaling/compare/4b06d09cc3545f5f386cf1d8fe7bd90165ff7a75..0bb2b59f8326a70c6438a1623f98d69ae4c4f34e).
2. Fixed a bug where we'd always use the "system" metrics port, even for LFC metrics: da48a4cd3f89ef8eaceaf620e7e6faff781aef77 (NOTE: this was from reading the code; I haven't yet tested it)
3. Switch from unnamed callbacks to a new `metricsMgr[M]` struct to contain them and the `kind` string: 5dc98a95675fcb6573278834a688b123eda89149
4. Switch from the `emptyMetrics()` constructor callback to a generics-based approach: 031cc2d13a74e4c232122053404b6119cae04f47
    - cc @Omrigan particularly want your feedback on this one. I could go either way on this -- left it as a separate commit so it's easy to revert.
5. Deduplicate metrics config parsing: 86cb9f4a277a5357f95920a82d25b074eef3e60a + c1ae8d3b1ed5c2badcf0d3873b939b6be70137a9

File: pkg/api/vminfo.go
Comment (Omrigan) [inline pkg/api/vminfo.go:403]:
```suggestion
func (c *ScalingConfig) validate(requireAll bool) error {
```

File: pkg/api/vminfo.go
Comment (Omrigan) [inline pkg/api/vminfo.go:380]: Shall we use `lo.ToPtr`?

File: pkg/agent/runner.go
Comment (Omrigan) [inline pkg/agent/runner.go:393]: > cc @Omrigan particularly want your feedback on this one. I could go either way on this -- left it as a separate commit so it's easy to revert.

Yeah, I'd revert it. It was educational - I learned `~` is a thing in golang :laughing:  But for real-life usage it'd stick to the trivial approach.

File: pkg/agent/runner.go
Comment (Omrigan) [inline pkg/agent/runner.go:451]: I know this is a prior change, but why this callback exists, and why is named `withLock`? Wouldn't just adding this `logger.Info` wherever `withLock` is called serve the same purpose, only simpler?

File: pkg/agent/core/metrics.go
Comment (Omrigan) [inline pkg/agent/core/metrics.go:91]: I'd just have 4 constants, but up to you.

File: pkg/agent/core/metrics.go
Comment (Omrigan) [inline pkg/agent/core/metrics.go:41]: WDYT about naming it `Metrics`, and using as a type parameter constraint instead of any? You mention "metric types" in the comment anyway, and this would help readability.

Also, two bottom lines are no longer true.

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go]: Haha fair enough- reverted in 60f835b5fc21bc867d8b3a466f1896a22d44f6dc

File: pkg/agent/core/metrics.go
Comment (sharnoff) [reply pkg/agent/core/metrics.go:38]: > Also, two bottom lines are no longer true.

Ah good catch, fixed in 662613d170363fa9cf240e5a0225e0c2f1c09674.

> and using as a type parameter constraint instead of any

Not quite sure I follow, could you elaborate?

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go]: Fair, yeah. Fixed in cb14dafe0dcefdb61bec02e365c9626af81b2cfa + b9e3ebd1ea51a7d3730b965d3ecd77c0318776f4 (split up to make squashing easier)

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go]: Added this as a new commit (to be kept separate after rebasing) in efdda7c819bbfe1a5178b884d9803438b5cebba1.

File: pkg/api/vminfo.go
Comment (sharnoff) [reply pkg/api/vminfo.go:380]: Ah yup, meant to do this - thanks for the reminder :) Implemented in #948 f9ed5d27408b77231d1e77d6338c6f248f939970

File: pkg/api/vminfo.go
Comment (sharnoff) [reply pkg/api/vminfo.go:403]: Fair enough. Implemented in #948 d77b97c5103c968bcee08496e8746daf9ae59027

**Comment by sharnoff** [comment]
Continuing with the changelog approach, if only because it helps me manage the rebasing:

1. Reverted 031cc2d13a74e4c232122053404b6119cae04f47 ("tmp: Switch from emptyMetrics() callback to generics"): 60f835b5fc21bc867d8b3a466f1896a22d44f6dc
2. Squashed everything back down: [`60f835b`..`666e11f`](https://github.com/neondatabase/autoscaling/compare/60f835b5fc21bc867d8b3a466f1896a22d44f6dc..666e11f02bae311c0ac57dbe129e42f1487262db)
3. Rebased onto main: 666e11f02bae311c0ac57dbe129e42f1487262db -> 34afd60dc06f5db3551ae1893054716a65cf5b2e
4. Renamed `newMetrics` -> `updateMetrics`: cb14dafe0dcefdb61bec02e365c9626af81b2cfa+b9e3ebd1ea51a7d3730b965d3ecd77c0318776f4
5. Fix outdated comment on `FromPrometheus`: 662613d170363fa9cf240e5a0225e0c2f1c09674
6. Renamed the `c`/`l` context/logger variables to `ctx2`/`logger2`: efdda7c819bbfe1a5178b884d9803438b5cebba1

Plus a couple more changes to #948 based on review here.

File: pkg/agent/runner.go
Comment (sharnoff) [reply pkg/agent/runner.go:450]: I think I did it at the time because it felt like a cleaner separation of concerns -- the caller is doing the logging anyways, so the executor stuff only needs to be concerned with doing the updating. 

Looking at it now, I'm not sure either way, and hesitant to change it. Happy to discuss more though.

File: pkg/agent/core/metrics.go
Comment (sharnoff) [reply pkg/agent/core/metrics.go]: Reworked this so we're the strings are directly passed to `getFloat`, similar to how LFC metrics parsing is now implemented, see 66b92893a804fc0f17b84f7c67358d9610573cb1.

File: pkg/agent/runner.go
Comment (Omrigan) [reply pkg/agent/runner.go:450]: > it's easier if that doesn't also have logging

It doesn't have logging, but if have to pass the callback, so not much profit :upside_down_face: 

> I'm not sure either way, and hesitant to change it.

No problem, that's was before the current change.

**Review by Omrigan — APPROVED** [review]
LGTM, pending comments and conflicts resolution.

**Comment by sharnoff** [comment]
Final changes:

1. Squash previous commits. Diff here (empty): [`66b9289`..`20bd2ba`](https://github.com/neondatabase/autoscaling/compare/66b92893a804fc0f17b84f7c67358d9610573cb1..20bd2ba55a40f0873971bb19d19d4ba8b9631704)
2. Rebase onto main (dropping #948 from this branch). Diff here: [`20bd2ba`..`fb2f910`](https://github.com/neondatabase/autoscaling/compare/20bd2ba55a40f0873971bb19d19d4ba8b9631704..fb2f9104ad9220f8d685f80d99554a1a480050bc)
3. Fix compile error from rebase: 689e18d0d66b935f9c0ffba6cae5df2e4f8f8306
4. Change `metricsMgr[M any]` -> `metricsMgr[M core.FromPrometheus]`: d0991a9d2447ff1c2a37e0cf2241422df51dea52

Waiting on some issues with latest release. Will hold off on merging for now to avoid accidentally adding much larger changes alongside a small fix.

File: pkg/agent/core/metrics.go
Comment (sharnoff) [reply pkg/agent/core/metrics.go:38]: Hm, that's fair. In rust, it's generally an anti-pattern to restrict generic types more than they need to be, but I'm not sure if it's the same in Go.

I like the idea of using it to signal purpose. Changed from `M any` to `M FromPrometheus` in d0991a9d2447ff1c2a37e0cf2241422df51dea52

**Pull request merged**
